### ISSUE TITLE
AWS: Problem recovering from bad mount

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#32630]
We had an error with a failed attach (in AWS), and I manually detached it.

But the AWS cloudprovider did not recover; it failed to recognize that the volume was already detached.

From KCM:

```
I0914 01:09:27.532272       5 reconciler.go:168] Started AttachVolume for volume "kubernetes.io/aws-ebs/aws://us-west-2a/vol-a0fb3428" to node "ip-172-20-0-206.us-west-2.compute.internal"
W0914 01:09:27.801863       5 aws.go:1195] Got assignment call for already-assigned volume: bm@vol-a0fb3428
I0914 01:09:27.874542       5 aws.go:1362] Waiting for volume state: actual=detached, desired=attached
I0914 01:09:28.910387       5 aws.go:1362] Waiting for volume state: actual=detached, desired=attached
I0914 01:09:30.874591       5 aws.go:1362] Waiting for volume state: actual=detached, desired=attached
I0914 01:09:31.914523       5 aws.go:1362] Waiting for volume state: actual=detached, desired=attached
```

**Comment by thomaso-mirodin** [comment]
Hey @justinsb do you know if logs reporting this behavior:
```
I0924 17:08:09.178959       1 reconciler.go:170] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0c87f7ef9d44df74c" to node "ip-10-20-11-34.us-west-2.compute.internal"
I0924 17:08:09.462097       1 aws.go:1140] Assigned mount device bd -> volume vol-0c87f7ef9d44df74c
E0924 17:09:14.441027       1 attacher.go:66] Error attaching volume "vol-0c87f7ef9d44df74c": Timeout waiting for volume state
E0924 17:09:14.441070       1 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/vol-0c87f7ef9d44df74c\"" failed. No retries permitted until 2016-09-24 17:11:14.441048379 +0000 UTC (durationBeforeRetry 2m0s). Error: AttachVolume.Attach failed for volume "kubernetes.io/aws-ebs/vol-0c87f7ef9d44df74c" (spec.Name: "storage") from node "ip-10-20-11-34.us-west-2.compute.internal" with: Timeout waiting for volume state
I0924 17:11:14.504587       1 reconciler.go:170] Started AttachVolume for volume "kubernetes.io/aws-ebs/vol-0c87f7ef9d44df74c" to node "ip-10-20-11-34.us-west-2.compute.internal"
W0924 17:11:14.712044       1 aws.go:1112] Got assignment call for already-assigned volume: bd@vol-0c87f7ef9d44df74c
E0924 17:12:18.694228       1 attacher.go:66] Error attaching volume "vol-0c87f7ef9d44df74c": Timeout waiting for volume state
E0924 17:12:18.694257       1 nestedpendingoperations.go:233] Operation for "\"kubernetes.io/aws-ebs/vol-0c87f7ef9d44df74c\"" failed. No retries permitted until 2016-09-24 17:14:18.694242912 +0000 UTC (durationBeforeRetry 2m0s). Error: AttachVolume.Attach failed for volume "kubernetes.io/aws-ebs/vol-0c87f7ef9d44df74c" (spec.Name: "storage") from node "ip-10-20-11-34.us-west-2.compute.internal" with: Timeout waiting for volume state
```

are likely caused by the same underlying issue you are fixing in #32635? Or do these seem like a different issue to you?

**Comment by justinsb** [comment]
This sounds like the issue we are fixing in the (coming soon) 1.3.8 and the (coming soon) 1.4.0.  I would probably go with one of those two versions when they come out.

Which version are you running @thomaso-mirodin

**Comment by thomaso-mirodin** [comment]
We're running a slight fork of 1.3.5 atm (adding some admission controllers):

```
docker exec -it 5ec8e58daa84 /hyperkube controller-manager --version
Kubernetes v1.3.5.1+e5f7974f7117ff
```

Is there another issue you can link me to? or are the fixes just generic volume manager fixes that have happened between 1.3.8 and 1.3.5?

**Comment by offlinehacker** [comment]
Hi, looks like we are having the same issue on 1.4.1

**Comment by thomaso-mirodin** [comment]
Hey @offlinehacker, our issue ended up being this one: https://github.com/kubernetes/kubernetes/issues/31891

What we are seeing appears to be a bug between with the Xen block device driver and the underlying hypervisor. We're talking to AWS support atm to see if they can assist.

**Comment by offlinehacker** [comment]
This is getting really annoying, our pods are not starting on dayli basis and then i have to spend a lot of time recovering them. While this is not critical for us as we have replicated services, it just takes a lot of time to recover. We are running it on k8s-1.4-debian-jessie-amd64-hvm-ebs-2016-09-30 with 4.4.23-k8s kernel. What do you do to workaround this issue, or have you managed to fix it?

**Comment by hanikesn** [comment]
Our problems with mounting AWS volumes actually got a lot worse when upgrading from 1.4.1 to 1.4.6. Now volumes don't get correctly mounted, when restarting a pod that's using an EBS volume.

**Comment by jingxu97** [comment]
@offlinehacker @hanikesn we have a few fixes recently related to AWS volumes #37302 and are working on back port the fix to 1.4. Please let me know if you think your issue is related to #37302 or not. 
If you prefer, I can also help check your log to see where the issues are. Sorry for the inconvenience.

**Comment by hanikesn** [comment]
We're running 1.6.2 now and didn't have any problems for a while, but we also reduced the number of PVCs that we're using quite a bit.
